django_mongodb_backend/indexes.py

```diff
@@ -101,6 +105,176 @@ def where_node_idx(self, compiler, connection):
+                    id=f"{self._error_id_prefix}.W001",
+                )
+            )
+        return errors
+
+    def search_index_data_types(self, field, db_type):
+        """
+        Map a model field's internal type to search index type.
+        Reference: https://www.mongodb.com/docs/atlas/atlas-search/define-field-mappings/#data-types
+        """
+        if field.get_internal_type() == "UUIDField":
+            return "uuid"
+        if field.get_internal_type() in {"ObjectIdAutoField", "ObjectIdField"}:
+            return "ObjectId"
+        if field.get_internal_type() == "EmbeddedModelField":
+            return "embeddedDocuments"
+        if db_type in {"int", "long"}:
+            return "number"
+        if db_type == "binData":
+            return "string"
+        if db_type == "bool":
+            return "boolean"
+        if db_type == "object":
+            return "document"
+        return db_type
+
+    def get_pymongo_index_model(
+        self, model, schema_editor, field=None, unique=False, column_prefix=""
+    ):
+        if not schema_editor.connection.features.supports_atlas_search:
+            return None
+        fields = {}
+        for field_name, _ in self.fields_orders:
+            field_ = model._meta.get_field(field_name)
+            type_ = self.search_index_data_types(field_, field_.db_type(schema_editor.connection))
+            field_path = column_prefix + model._meta.get_field(field_name).column
+            fields[field_path] = {"type": type_}
+        return SearchIndexModel(
+            definition={"mappings": {"dynamic": False, "fields": fields}}, name=self.name
+        )
+
+
+class VectorSearchIndex(SearchIndex):
+    suffix = "vsi"
+    VALID_SIMILARITIES = frozenset(("euclidean", "cosine", "dotProduct"))
+    _error_id_prefix = "django_mongodb_backend.indexes.VectorSearchIndex"
+
+    def __init__(self, *expressions, fields=(), similarities="cosine", name=None, **kwargs):
+        super().__init__(*expressions, fields=fields, name=name, **kwargs)
+        self.similarities = similarities
+        self._multiple_similarities = isinstance(similarities, tuple | list)
+        for func in similarities if self._multiple_similarities else (similarities,):
+            if func not in self.VALID_SIMILARITIES:
+                raise ValueError(
+                    f"'{func}' isn't a valid similarity function "
+                    f"({', '.join(sorted(self.VALID_SIMILARITIES))})."
+                )
+        seen_fields = set()
+        for field_name, _ in self.fields_orders:
+            if field_name in seen_fields:
+                raise ValueError(f"Field '{field_name}' is duplicated in fields.")
+            seen_fields.add(field_name)
+
+    def check(self, model, connection):
+        errors = super().check(model, connection)
+        expected_similarities = 0
+        for field_name, _ in self.fields_orders:
+            field_ = model._meta.get_field(field_name)
```

Comment:
I think the underscore suffix on "field" is an artifact of a previous location of these checks. As far as I see, it's not needed.